docs/user-guide/setting-up-ci-cd.rst

```diff
@@ -0,0 +1,165 @@
+data validate``, either in a seperate check or as part of the data validation
+check.
+
+Train a Model
+#############
+
+Training a model verifies that your NLU pipeline and policy configurations are
+valid and trainable, and it provides a model to test on end-to-end test
+stories. Training a model is also :ref:`part of the continuous deployment
+process <uploading-a-model>`, as you'll need a model to upload to your server. 
+
+End-to-End Testing
+##################
+
+Testing your trained model on :ref:`end-to-end test stories
+<end-to-end-testing>` is the best way to have confidence in how your assistant
+will act in certain situations. These stories, written in a modified story
+format, allow you to provide entire conversations and test that, given this
+user input, your model will behave in the expected manner. This is especially
+important as you start introducing more complicated stories from user
+conversations. End-to-end testing is only as thorough and accurate as the test
+cases you write, so you should always update your end-to-end stories
+concurrently with your training stories.
+
+Note: End-to-end stories do **not** execute your action code. You will need to
+test your action code in a seperate step.
+
+NLU Comparison
+##############
+
+If you've made significant changes to your NLU training data (such as adding or
+splitting intents, or just adding/changing a lot of examples), you should run a
+:ref:`full NLU evaluation <evaluating-an-nlu-model>`. You'll want to compare
+the performance of the NLU model without your changes to an NLU model with your
+changes. You can do this by running NLU testing in cross-validation mode, or by
+training a model on a training set and testing it on a test set. If you use the
+latter approach, it is best to shuffle and split your data every time, as
+opposed to using a static NLU test set, which can easily become outdated. Since
+this can be a fairly resource intensive test, you can set this test to run only
+when a certain tag (e.g. "NLU testing required") is present, or only when
+changes to NLU data or the NLU pipeline were made.
+
+Testing Action Code
+###################
+
+The approach used to test your action code will depend on how it is
+implemented. Whichever method of testing your code you choose, you should
+include running those tests in your CI pipeline as well. 
+
+Continuous Deployment
+---------------------
+
+To get improvements out to your users frequently, you need to automate as
+much of the deployment process as possible. 
+
+CD steps usually run on push or merge to a certain branch, once CI checks have
+succeeded.
+
+.. contents::
+   :local:
+
+.. _uploading-a-model:
+
+Deploying your Rasa Model
+#########################
+
+You should already have a trained model from running end-to-end testing in your
+CI pipeline. You can set up your pipeline to upload the trained model to your
+Rasa server. If you're using Rasa X, you can also 
```

Comment:
Lets show them API calls for those (upload model in rasa x at least)

Reply:
are these right for the general case?

Reply:
Are these right for the general case?

Reply:
(the ones I added)